libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp

```diff
@@ -0,0 +1,560 @@
+    return true;
+}
+
+bool AP_ExternalAHRS::get_origin(Location &loc)
+{
+    WITH_SEMAPHORE(sem);
+    loc = origin;
+    return origin_set;
+}
+
+bool AP_ExternalAHRS::get_location(Location &loc)
+{
+    if (!last_pkt2) {
+        return false;
+    }
+    const struct VN_packet2 &pkt2 = *last_pkt2;
+    WITH_SEMAPHORE(sem);
+    loc = Location();
+    loc.lat = pkt2.GPS1posLLA[0] * 1.0e7;
+    loc.lng = pkt2.GPS1posLLA[1] * 1.0e7;
+    loc.alt = pkt2.GPS1posLLA[2] * 1.0e2;
+    return true;
+}
+
+Vector2f AP_ExternalAHRS::get_groundspeed_vector()
+{
+    if (!last_pkt1) {
+        return Vector2f();
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    return Vector2f(pkt1.velNED[0], pkt1.velNED[1]);
+}
+
+bool AP_ExternalAHRS::get_velocity_NED(Vector3f &vel)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    vel.x = pkt1.velNED[0];
+    vel.y = pkt1.velNED[1];
+    vel.z = pkt1.velNED[2];
+    return true;
+}
+
+bool AP_ExternalAHRS::get_speed_down(float &speedD)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    speedD = pkt1.velNED[2];
+    return true;
+}
+
+bool AP_ExternalAHRS::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    if (!healthy()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "VectorNav unhealthy");
+        return false;
+    }
+    if (last_pkt2->GPS1Fix < 3) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "VectorNav no GPS1 lock");
+        return false;
+    }
+    if (last_pkt2->GPS1Fix < 3) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "VectorNav no GPS2 lock");
+        return false;
+    }
+    return true;
+}
+
+/*
+  get filter status. We don't know the meaning of the status bits yet,
+  so assume all OK if we have GPS lock
+ */
+void AP_ExternalAHRS::get_filter_status(nav_filter_status &status) const
+{
+    memset(&status, 0, sizeof(status));
+    if (last_pkt1 && last_pkt2) {
+        status.flags.initalized = 1;
+    }
+    if (healthy() && last_pkt2) {
+        status.flags.attitude = 1;
+        status.flags.vert_vel = 1;
+        status.flags.vert_pos = 1;
+
+        const struct VN_packet2 &pkt2 = *last_pkt2;
+        if (pkt2.GPS1Fix >= 3) {
+            status.flags.horiz_vel = 1;
+            status.flags.horiz_pos_rel = 1;
+            status.flags.horiz_pos_abs = 1;
+            status.flags.pred_horiz_pos_rel = 1;
+            status.flags.pred_horiz_pos_abs = 1;
+            status.flags.using_gps = 1;
+        }
+    }
+}
+
+Vector3f AP_ExternalAHRS::get_gyro(void)
+{
+    if (!last_pkt1) {
+        return Vector3f();
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    return Vector3f(pkt1.gyro[0], pkt1.gyro[1], pkt1.gyro[2]);
```

Comment:
```suggestion
    return Vector3f{pkt1.gyro[0], pkt1.gyro[1], pkt1.gyro[2]};
```

This PR is showing the value of having a constructor on `Vector3f` which takes a `float[3]`.  The equivalent exists on `VectorN`.

Would save the creation of the reference in this case to as you'd just

```suggestion
    return Vector3f{last_pkt1->gyro};
```


... would save a chunk of code, actually, as the creation is done with a memcpy.

Reply:
possibly more error prone, as constructor would need to be for just a pointer, with no length tag